p4-16/spec/P4-16-spec.mdk

```diff
@@ -6607,11 +6629,269 @@ control ingress(inout Header_t h, inout Meta_t m,
+
+The priorities determined in this way are the values used when the P4
+program is first loaded into a device.  Afterwards, the priorities may
+only change by means provided by the control plane API in use.
+
+In the second step, the compiler issues errors for out of range
+priority values, and/or warnings for certain combinations of entry
+priorities that might be unintended by the developer, unless the
+developer explicitly disables those warnings.
+
+If any priority values are negative, or larger than the maximum
+supported value, that is a compile time error.
+
+If the annotation `@noWarn("duplicate_priorities")` is _not_ used on
+the `initial_entries` table property, then the compiler issues a
+warning if any two entries for the same table have equal priority
+values.  Both P4Runtime and TDI leave it unspecified which entry is
+the winner if a lookup key matches multiple keys that all have the
+same priority, hence a warning is useful to less experienced
+developers that are unfamiliar with this unspecified behavior.
+
+If the annotation `@noWarn("duplicate_priorities")` is used on the
+`initial_entries` table property, then no warnings of this type are
+ever issued by the compiler.  Using equal priority values for multiple
+entries in the same table is sometimes useful in reducing the number
+of hardware updates required when adding entries to such a table.
+
+If the annotation `@noWarn("entries_out_of_priority_order")` is _not_
+used on the `initial_entries` table property, then the compiler issues
+a warning if:
+
+- If `largest_priority_wins` is `true` for the table, and there is any
+  pair of consecutive entries where `prio[j] < prio[j+1]`, then a
+  warning is issued for that pair of entries.
+- If `largest_priority_wins` is `false` for the table, and there is any
+  pair of consecutive entries where `prio[j] > prio[j+1]`, then a
+  warning is issued for that pair of entries.
+
+This warning is useful to developers that want the order that entries
+appear in the source code to match the relative priority of entries in
+the target device.
+
+If the annotation `@noWarn("entries_out_of_priority_order")` is used
+on the `initial_entries` table property, then no warnings of this type
+are ever issued by the compiler for this table.  This option is
+provided for developers who explicitly choose to specify entries in an
+order that does not match their relative priority order.
+
+The following example is the same as the first example in section
+[#sec-entries], except for the definition of table `t_exact_ternary`
+shown below.
+
+~ Begin P4Example
+table t_exact_ternary {
+    key = {
+        h.h.e : exact;
+        h.h.t : ternary;
+    }
+
+    actions = {
+        a;
+        a_params;
+    }
+
+    default_action = a;
+
+    largest_priority_wins = false;
+    priority_delta = 10;
+    @noWarn("duplicate_priorities")
+    initial_entries = {
+        const 10: { (0x01, 0x1111 &&& 0xF   ), a_params(1) };
```

Comment:
I find this syntax hard to read. I really liked the explicit "priority"

Reply:
The grammar allows an optional `priority=` before the number, if you want, but does not require it.

I'm at the point where as soon as we find a syntax that bison is happy with and most people reviewing it are happy with, I'm good to go :-)